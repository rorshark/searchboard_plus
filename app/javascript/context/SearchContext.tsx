import React from 'react'
import { useSearch, SearchAPI, SearchParams } from '../hooks/useSearch'

export const defaultSearchApi: SearchAPI = {
  error: false,
  loading: true,
  customers: [],
  search: '',
  setSearch: (_) => {},
  companies: [],
  selectedCompany: null,
  setSelectedCompany: (_) => {},
  fetchCustomers: async () => {}
}

export const SearchContext: React.Context<SearchAPI> = React.createContext(defaultSearchApi)

export const SearchProvider: React.FC<{
  initialSearchParams: SearchParams
}> = ({ initialSearchParams, children }) => {
  const search = useSearch(initialSearchParams)

  return (
    <SearchContext.Provider value={search}>
      {children}
    </SearchContext.Provider>
  )
}
