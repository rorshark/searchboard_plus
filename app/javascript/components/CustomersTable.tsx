import React, { useContext } from 'react'
import { SearchContext } from '../context/SearchContext'

const CustomersTable: React.FC = () => {
  const { error, loading, customers } = useContext(SearchContext)

  if (error) return (
    <div data-testid="error">
      Oops, something went wrong.
    </div>
  )

  if (loading) return (
    <div data-testid="loading">
      Loading...
    </div>
  )

  return (
    <table data-testid="customers-table">
      <thead>
        <tr>
          <th>First Name</th>
          <th>Last Name</th>
          <th>Company Name</th>
        </tr>
      </thead>
      <tbody>
        {customers.map((customer, index) => (
          <tr key={index}>
            <td>{customer.firstName}</td>
            <td>{customer.lastName}</td>
            <td>{customer.company?.companyName}</td>
          </tr>
        ))}
      </tbody>
    </table>
  )
}

export default CustomersTable
