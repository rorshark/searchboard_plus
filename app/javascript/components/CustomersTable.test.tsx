import React from 'react'
import { render, screen } from '@testing-library/react'
import CustomersTable from './CustomersTable'
import { SearchContext } from '../context/SearchContext'

test('it renders a loading screen', () => {
  const search = {
    loading: true,
    customers: [],
    error: false
  }

  render(
    <SearchContext.Provider value={search}>
      <CustomersTable />
    </SearchContext.Provider>
  )

  const loading = screen.getByTestId('loading')
  const error = screen.queryByTestId('error')
  const table = screen.queryByTestId('customers-table')

  expect(loading).toMatchSnapshot()
  expect(error).toBeNull()
  expect(table).toBeNull()
})

test('it renders an error screen', () => {
  const search = {
    loading: false,
    customers: [],
    error: true
  }

  render(
    <SearchContext.Provider value={search}>
      <CustomersTable />
    </SearchContext.Provider>
  )

  const error = screen.getByTestId('error')
  const loading = screen.queryByTestId('loading')
  const table = screen.queryByTestId('customers-table')

  expect(error).toMatchSnapshot()
  expect(loading).toBeNull()
  expect(table).toBeNull()
})

test('it renders the customers table when loaded', async () => {
  const search = {
    loading: false,
    customers: [{ firstName: 'Dougy', lastName: 'Pauly' }],
    error: false
  }

  render(
    <SearchContext.Provider value={search}>
      <CustomersTable />
    </SearchContext.Provider>
  )

  const table = screen.getByTestId('customers-table')
  const error = screen.queryByTestId('error')
  const loading = screen.queryByTestId('loading')

  expect(table).toMatchSnapshot()
  expect(error).toBeNull()
  expect(loading).toBeNull()
})
