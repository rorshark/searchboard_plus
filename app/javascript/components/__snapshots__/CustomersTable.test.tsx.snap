// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a loading screen 1`] = `
<div
  data-testid="loading"
>
  Loading...
</div>
`;

exports[`it renders an error screen 1`] = `
<div
  data-testid="error"
>
  Oops, something went wrong.
</div>
`;

exports[`it renders the customers table when loaded 1`] = `
<table
  data-testid="customers-table"
>
  <thead>
    <tr>
      <th>
        First Name
      </th>
      <th>
        Last Name
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Dougy
      </td>
      <td>
        Pauly
      </td>
    </tr>
  </tbody>
</table>
`;
